File: myurlshortner-consumer/myurlshortner-consumer/src/main/scala/com/acme/myurlshortner/consumer/application/usecase/ShortenedUrlUserEventsUseCases.scala
package com.acme.myurlshortner.consumer.application.usecase

import com.acme.events.UserAccessedShortenedUrl
import com.acme.myurlshortner.consumer.domain.command.SaveShortenedUrlUserAccessCommand
import com.acme.myurlshortner.consumer.domain.service.ShortenedUrlUserEventsService
import zio._

import java.net.URI

class ShortenedUrlUserEventsUseCases(
  val service: ShortenedUrlUserEventsService
) {

  def handleUserAccessedShortenedUrl(event: UserAccessedShortenedUrl): ZIO[Any, Throwable, Unit] = for {
    originalUrl  <- ZIO.attempt(URI.create(event.original_url.toString()))
    shortenedUrl <- ZIO.attempt(URI.create(event.shortened_url.toString()))
    userAgent    <- ZIO.succeed(event.user_agent.toString())
    _            <- service.saveUserAccess(
                      SaveShortenedUrlUserAccessCommand(userAgent, originalUrl, shortenedUrl)
                    )
  } yield (ZIO.unit)
}

object ShortenedUrlUserEventsUseCases {
  val layer: ZLayer[ShortenedUrlUserEventsService, Nothing, ShortenedUrlUserEventsUseCases] = ZLayer.fromZIO(
    for {
      service <- ZIO.service[ShortenedUrlUserEventsService]
    } yield (ShortenedUrlUserEventsUseCases(service))
  )
}

File: myurlshortner-consumer/src/main/kotlin/com/acme/myurlshortner/consumer/application/usecase/ShortenedUrlUserEventsUseCases.kt
package com.acme.myurlshortner.consumer.application.usecase

import com.acme.events.UserAccessedShortenedUrl
import com.acme.events.UserCreatedShortenedUrl
import com.acme.myurlshortner.consumer.domain.userevent.command.UserAccessedShortenedUrlCommand
import com.acme.myurlshortner.consumer.domain.userevent.command.UserCreatedShortenedUrlCommand
import com.acme.myurlshortner.consumer.domain.userevent.service.UserAccessedShortenedUrlEventService
import com.acme.myurlshortner.consumer.domain.userevent.service.UserCreatedShortenedUrlService
import org.springframework.stereotype.Component
import java.net.URI
import java.time.OffsetDateTime

@Component
class ShortenedUrlUserEventsUseCases(
    private val userAccessEventService: UserAccessedShortenedUrlEventService,
    private val userShortenedUrlService: UserCreatedShortenedUrlService
) {

    fun handleUserAccessedShortenedUrl(event: UserAccessedShortenedUrl) {
        val command = UserAccessedShortenedUrlCommand(
            uniqueIdentifier = event.uniqueIdentifier,
            originalUrl = URI.create(event.originalUrl),
            shortenedUrl = URI.create(event.shortenedUrl),
            userAgent = event.userAgent,
            accessedAt = event.accessedAt?.let { OffsetDateTime.parse(it) } ?: OffsetDateTime.now()
        )
        userAccessEventService.handleShortenedUrlUserAccessed(command)
    }

    fun handleUserCreatedShortenedUrl(event: UserCreatedShortenedUrl) {
        userShortenedUrlService.handleShortenedUrlCreated(
            UserCreatedShortenedUrlCommand(
                uniqueIdentifier = event.uniqueIdentifier,
                originalUrl = URI.create(event.originalUrl),
                userId = event.userId,
                title = event.title,
                createdAt = OffsetDateTime.parse(event.createdAt)
            )
        )
    }
}
